src/server/node_manager.c

```diff
@@ -2134,95 +2128,122 @@ stat_update(int stream)
 
-	while (njobs--) {
-		struct resc_used_update *prused;
+	log_eventf(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG, __func__, "received obits = %d", njobs);
 
-		/* IMPORTANT NOTE					      */
-		/* allocate resc_used_update here, but leave it to job_obit() */
-		/* to free it when job_obit() has completed		      */
+	reject_list = (char **)calloc(1, njobs * sizeof(char *));
+	if (reject_list == NULL)
+		goto recv_job_obit_err;
 
-		prused = (struct resc_used_update *)
-			malloc(sizeof(struct resc_used_update));
-		if (prused != 0) {
+	ack_list = (char **)calloc(1, njobs * sizeof(char *));
+	if (ack_list == NULL)
+		goto recv_job_obit_err;
 
-			prused->ru_next   = NULL;
-			prused->ru_pjobid = NULL;
+	while (njobs--) {
+		CLEAR_HEAD(rused.ru_attr);
+		rused.ru_comment = NULL;
+		rused.ru_next = NULL;
+		rused.ru_pjobid = NULL;
 
-			if (decode_stat_update(stream, prused) == 0) {
+		if (decode_stat_update(stream, &rused) == 0) {
+			int is_reject = 0;
+
+			DBPRT(("recv_job_obit: decoded obit for %s\n", rused.ru_pjobid))
+			job_obit(&rused, stream, &is_reject);
+			if (is_reject == 1) {
+				reject_list[reject_count++] = rused.ru_pjobid;
+				rused.ru_pjobid = NULL;
+			} else if (is_reject != -1) { /* -1 means ignore ruu */
+				ack_list[ack_count++] = rused.ru_pjobid;
+				rused.ru_pjobid = NULL;
+			}
+			free(rused.ru_comment);
+			if (rused.ru_pjobid != NULL)
+				free(rused.ru_pjobid);
+			free_attrlist(&rused.ru_attr);
+		} else
+			goto recv_job_obit_err;
+	}
 
-				DBPRT(("recv_job_obit: decoded obit for %s\n",
-					prused->ru_pjobid))
-				job_obit(prused, stream);
-			} else {
-				mominfo_t *mp;
+	log_eventf(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG, __func__, "processed obits, sending replies acks: %d, rejects: %d", ack_count, reject_count);
 
-				DBPRT(("recv_job_obit: failed to decode obit for %s\n",
-					prused->ru_pjobid))
-				/* had a error, discard rest of message */
-				if ((mp = tfind2((u_long)stream, 0, &streams)) != NULL) {
-					log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_NODE,
-						LOG_NOTICE, mp->mi_host, "error in recv_job_obit");
-				}
-				tpp_eom(stream);
-				FREE_RUU(prused)
-			}
+	if (ack_count > 0) {
+		if (is_compose(stream, IS_ACKOBIT) != DIS_SUCCESS)
+			goto recv_job_obit_err;
+		if (diswui(stream, ack_count) != DIS_SUCCESS)
+			goto recv_job_obit_err;
+		for (i = 0; i < ack_count; i++) {
+			if (diswst(stream, ack_list[i]) != DIS_SUCCESS)
+				goto recv_job_obit_err;
+			free(ack_list[i]);
+			ack_list[i] = NULL;
+		}
+		dis_flush(stream);
+	}
+	free(ack_list);
+	ack_list = NULL;
+	ack_count = 0;
+
+	if (reject_count > 0) {
+		if (is_compose(stream, IS_BADOBIT) != DIS_SUCCESS)
+			goto recv_job_obit_err;
+		if (diswui(stream, reject_count) != DIS_SUCCESS)
+			goto recv_job_obit_err;
+		for (i = 0; i < reject_count; i++) {
+			if (diswst(stream, reject_list[i]) != DIS_SUCCESS)
+				goto recv_job_obit_err;
+			free(reject_list[i]);
+			reject_list[i] = NULL;
 		}
+		dis_flush(stream);
```

Comment:
instead of sending two different replies, can we send only one?
the reply will be job_id followed by 0 or 1 indicating if its is a bad obit or not.
in the MoM, we introduce a new message type like IS_OBITREPLY